### ISSUE TITLE
Tracking issue for eRFC 2318, Custom test frameworks

### ISSUE THREAD

**Issue by Centril** [rust-lang/rust#50297]
This is a tracking issue for the eRFC "Custom test frameworks" (rust-lang/rfcs#2318).

**Documentation:**
* Implementation proposal: https://www.jrenner.net/rust/testing/2018/08/06/custom-test-framework-prop.html
* https://doc.rust-lang.org/nightly/unstable-book/language-features/custom-test-frameworks.html

**Steps:**

- [ ] Implement the RFC (cc @rust-lang/dev-tools @rust-lang/cargo -- can anyone write up mentoring instructions?)
- [ ] Adjust documentation ([see instructions on forge][doc-guide])
- [ ] Stabilization PR ([see instructions on forge][stabilization-guide])

[stabilization-guide]: https://forge.rust-lang.org/stabilization-guide.html
[doc-guide]: https://forge.rust-lang.org/stabilization-guide.html#updating-documentation

**Unresolved questions:**

- [ ] [Integration with doctests](https://github.com/rust-lang/rfcs/blob/master/text/2318-custom-test-frameworks.md#integration-with-doctests)
- [ ] [Standardizing the output](https://github.com/rust-lang/rfcs/blob/master/text/2318-custom-test-frameworks.md#standardizing-the-output-1)
- [ ] [Namespacing](https://github.com/rust-lang/rfcs/blob/master/text/2318-custom-test-frameworks.md#namespacing)
- [ ] [Runtime dependencies and flags](https://github.com/rust-lang/rfcs/blob/master/text/2318-custom-test-frameworks.md#runtime-dependencies-and-flags)
- [ ] [Naming](https://github.com/rust-lang/rfcs/blob/master/text/2318-custom-test-frameworks.md#naming)
- [ ] [Bencher](https://github.com/rust-lang/rfcs/blob/master/text/2318-custom-test-frameworks.md#bencher)
- [ ] [Cargo integration](https://github.com/rust-lang/rfcs/blob/master/text/2318-custom-test-frameworks.md#cargo-integration)

**Notes:**

- Destabilization of `#[bench]`: https://github.com/rust-lang/rust/issues/63798#issuecomment-526593504

**Implementation history:**
* Initial implementation: #53410

**Comment by CAD97** [comment]
On stdout/err capture (after doing some research around the proof-of-concept #50457):

The current state is sub-optimal, to say the least. Both `stdout` and `stderr` exist as thread-local handles that only the stdlib has access to, which `print` family of macros use, as does `panic`.

<details><summary>Relevant libstd snippets</summary>

https://github.com/rust-lang/rust/blob/935a2f12b2a3ce44cbcbd0f7de1b388949cccca2/src/libstd/io/stdio.rs#L23-L27
https://github.com/rust-lang/rust/blob/935a2f12b2a3ce44cbcbd0f7de1b388949cccca2/src/libstd/panicking.rs#L38-L42
https://github.com/rust-lang/rust/blob/935a2f12b2a3ce44cbcbd0f7de1b388949cccca2/src/libstd/io/stdio.rs#L614-L711
</details><br>

The existing stdout/err capture for the test harness works by setting the thread-local, which, if set, the `print` family of macros will use, otherwise falling back to `io::stdout()`. This setup is problematic because it's fairly trivial to escape this (#12309) and 99.9% of logging frameworks will escape accidentally (#40298).

The reason for this is that any solution that's generic over `io::Write` is going to use `io::stdout()` as the sink to hit stdout, and that doesn't go through the tread-local plumbing, so hits the true stdout directly.

The way I see it, there are three ways to resolve this quirk:

- Create a new stdlib `io::Write` sink that uses the thread-local plumbing that the `print` macros do today, and implement the `print` macros as writing to that sink. That's the approach I proved was possible with #50457.
- Create a new `io::Write` sink that uses (`e`)`print!` behind the scenes. This is possible entirely in an external crate:

<details><summary>(<code>E</code>)<code>PrintWriter</code> - feel free to use these, I am the author and release them under MIT/Unlicense</summary>

```rust
/// IO sink that uses the `print` macro, for test stdout capture.
pub struct PrintWriter;
impl std::io::Write for PrintWriter {
    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
        print!("{}", String::from_utf8_lossy(buf));
        Ok(buf.len())
    }
    fn flush(&mut self) -> std::io::Result<()> { std::io::stdout().flush() }
}

/// IO sink that uses the `eprint` macro, for test stdout capture.
pub struct EPrintWriter;
impl std::io::Write for EPrintWriter {
    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
        eprint!("{}", String::from_utf8_lossy(buf));
        Ok(buf.len())
    }
    fn flush(&mut self) -> std::io::Result<()> { std::io::stdout().flush() }
}
```
</details><br>

Unfortunately, this suffers from a large number of drawbacks. Because we only have access to the `print` macro, we have to interpret whatever sequence of bytes the sink gets as a UTF-8 string, even though that's potentially not true, etc. etc. not zero-cost when it's not, and so on. It's a decent solution for hooking up logging in tests, but is very much not a general solution that could be used outside tests.

That takes us to the third option, which is almost certainly the best, but also the most complicated:

- Actually change stdout/err

It sounds simple, but it's a lot simpler said than done with the _current_ test harness setup. `io::Stderr` boils down to a handle to a write stream available to the process, and is a _process-local_ resource -- there's only _one_ stdout per process. The test harness currently runs tests parallel on different threads within the same process, so

**Comment by CAD97** [comment]
TL;DR:

To have sane capture behavior for stdout and stderr for `print`-family macros, `panic`, and `io::{stdout, stderr}`, running each test in its own (child) process is probably required. I'd love to help out wherever I can, but lack confidence.

**Comment by retep998** [comment]
>The only solution I see at this time that maintains current behavior of separate captures per test run is to run them all in their own process, and capture stdout that way.

This would be significantly slower on certain platforms where process creation is slow, notably Windows. Especially when the tests are all very quick this will cause severe regressions in total time.

**Comment by CAD97** [comment]
Another possibility for process separation of tests is to spawn a process for each unit of parallelism, but each process is reused for multiple tests.

Of course, we could also just promote the thread-local replaceable stdout/err to `io::Stdout`/`Err`, but that would be suboptimal itself.

The issue is that stdout is, as far as I can tell, a process-local resource. To capture `io::Stdout` we either need our own stdout replacement machinery on top or to separate processes.

**Comment by SoniEx2** [comment]
if I want to argue about this do I argue about it here?

so like I made `hexchat-plugin` crate. the problem is you can't test hexchat side without hexchat. so I'd need some sort of `#[hexchat_test]`.

but I still want the users to be able to use plain old vanilla rust `#[test]`s. I shouldn't have to reimplement/replace the whole testing system just to be able to produce some hexchat-based unit tests (in addition to normal tests).

(disclaimer: I don't fully understand the RFC, but it seems to require either replacing everything, or not having tests)

**Comment by jonhoo** [comment]
@SoniEx2 this RFC is about how we'd like to change the underlying mechanism that is used to arrange testing in Rust so that authors can write new test frameworks (e.g., quickcheck) that have access to the same mechanisms as existing built-in tests ( able to access private functions, can find all tagged functions, etc.). The current proposal is relatively wholesale (you change the test framework for a whole crate), but this is something that may [be improved upon](https://github.com/rust-lang/rfcs/blob/master/text/2318-custom-test-frameworks.md#cargo-integration-1). We decided against more sophisticated designs in the initial implementation phase (this is just an "experimental" RFC) so that we don't bite off more than we can chew, but I think down the line we'd want more fine-grained test framework selection too.

As for your specific example, you don't actually need custom frameworks for that feature :) It seems like you want to only run a test if hexchat isn't running? To do that, you could either mark the test with `#[allow_fail]`, or you could have the test just `return` if it fails to connect. If you wanted a more fine-grained approach, you could write a procedural macro that re-writes any function annotated with `#[hexchat_test]` into a `#[test]` that returns if hexchat isn't running.

**Comment by SoniEx2** [comment]
no you don't understand, they have to run inside hexchat. loaded with dlopen (or equivalent). after plugin registration.

**Comment by djrenren** [comment]
Hey all, I've come up with a less-heavyweight proposal and implementation:

Proposal: https://blog.jrenner.net/rust/testing/2018/08/06/custom-test-framework-prop.html
Implementation: https://github.com/djrenren/rust/tree/custom-test-frameworks

**Comment by djrenren** [comment]
Hey all, I've come up with a simplified proposal and I have an implementation now (though it needs a little cleaning up before a PR). 

https://blog.jrenner.net/rust/testing/2018/08/06/custom-test-framework-prop.html

Lemme know what you think!

**Comment by jonhoo** [comment]
@djrenren finally had a chance to read your proposal. I'm excited by the progress on this, though your write-up did raise some questions for me:

 - Early on, you say that "Annotated items must be a nameable `const`, `static`, or `fn`.". Further down, when you discuss the use-case of a framework that supports test suites, you show a `#[test_suite]` annotation on a `mod` with `#[suite_member]` on the nested tests. It'd be good to show how that expansion works. Does a proc macro turn `#[test_suite]` into a `const` with an element for each `#[suite_member]`? How does it find all of them? Are the `#[suite_member]`s all turned into `#[test_case]`? How do you avoid double-executing those test cases (once as a "child" and once due to the `#[test-case]` annotation)?
 - For `#![test_runner]` you say "The type of the function must be `Fn(&[&mut Foo]) -> impl Termination` for some `Foo` which is the test type", yet further down you write a `Fn(&[&dyn Foo])`. I assume the latter is the correct version?
 - Is `TestDescAndFn` sufficient for our purposes here? I don't think that can (currently) wrap a `struct` or `const` in a meaningful way?
 - I find the "test runner" and "test format" examples a little backwards. To me, "formatting" is about output, and "running" is about the structure of the tests themselves (and how they're run). We can bikeshed names later though.
 - The "editor wants to integrate with testing" example makes me sad. It seems really sad for the crate-under-test to have to be modified to work with my editor. One of the earlier RFCs had a section on separating the test harness from the test executor, with the former controlling output and the latter how tests are written and executed. And crucially, the former should be possible to swap *without changing the source code of the c-u-t*. That doesn't solve the case of an editor wanting to control test execution as well, but I think ideally we'd be able to fold that into the harness somehow too.

**Comment by CAD97** [comment]
I think in most cases the CUT won't need to change itself to integrate with the IDE test runner, so long as the IDE understands the test runner you're working with. As I understand how JetBrains' IntelliJ IDEA integrates with both JUnit and Rust's libtest is that it communicates using the standard API and stdin/out. JUnit's integration is obviously tighter, as IDEA was a Java IDE first, but the integration with libtest tests in Rust is already quite good even without any specific integration required.

Obviously the IDE needs to understand how the test runner works or the test runner needs to understand how the IDE works, but under most cases of using a popular test runner, the CUT shouldn't need to change to be tested with in-IDE functionality.

**Comment by SoniEx2** [comment]
I'm gonna assume this isn't for me, as a crate author, but for IDE developers.

Let me know when crate authors are part of the plan.

**Comment by djrenren** [comment]
@jonhoo 

> Early on, you say that...

I don't want to get too bikesheddy on how that particular macro would work but basically we'd just produce a single const that had the `#[test_case]` annotation. The children would be aggregated inside it. 

> I assume the latter is the correct version?

`Fn(&[&Foo]) -> ()` is a subtype `Fn(&[&mut Foo]) -> impl Termination`. It's not required that Foo be a Trait but if it is, then you need `dyn`, so those two lines aren't in contradiction.

> Is TestDescAndFn sufficient for our purposes here?

I agree `TestDescAndFn` is not sufficient, though it could successfully be constructed from all sorts of things with the use of proc macros. I think for libtest we should just make a trait that essentially mirrors `TestDescAndFn` and accept that.

> One of the earlier RFCs had a section on separating the test harness from the test executor, with the former controlling output and the latter how tests are written and executed.

This is still entirely possible under the current design, it just requires the test runner to be designed to be pluggable.

**Comment by djrenren** [comment]
@SoniEx2 It is for crate authors! 😄 

For most crate authors (unless you're authoring test machinery), you'll just import any macros you want for declaring tests or benchmarks, and then you'll pick a test runner. For most cases the built-in one should be fine though.

**Comment by SoniEx2** [comment]
but how do you run tests inside a hexchat instance, considering the whole hexchat plugin loading system and stuff?

**Comment by SoniEx2** [comment]
I mean you can't (easily) run hexchat (a GUI-based networked program) in cargo, can you?

**Comment by SoniEx2** [comment]
can I assume this won't help test integrations (how your plugin behaves running inside a host program) and is more geared towards testing your plugin's independent parts?

**Comment by djrenren** [comment]
For those worried about stdout capture, I've just demonstrated how this proposal can allow for @CAD97's style of multiprocess stdout capturing:
https://github.com/djrenren/rust-test-frameworks/blob/master/multiprocess_runner/src/lib.rs

This code compiles and runs under my PR: https://github.com/rust-lang/rust/pull/53410

The output of `cargo test` is available in the same directory and is called `output.txt`

**Comment by CAD97** [comment]
And just for clarity, the example does spawn a process for each test, [which is suboptimal on OSes where process creation is slow, e.g. Windows](https://github.com/rust-lang/rust/issues/50297#issuecomment-389059878). A real runner would implement a scheme like I described earlier to avoid spawning more processes than necessary. (And actually parallelize the running of tests.)

**Comment by japaric** [comment]
Whoops, I'm (very) late to the party.

I have read the blog post but not looked at the rustc implementation and I'm concerned with the presence of `impl Termination` and `String` (see `Testable.name`) in the proposal.

How will this proposal work with the version of `#![no_std]` binaries that's being stabilized for the edition release? Those `#![no_std]` binaries use `#![no_main]` to avoid the unstable `start` (i.e. `fn main()` interface) and `termination` (i.e. `Termination` trait) lang items. Without `start` if `rustc --test` injects a `main` function into the crate it will end up being discarded as these binaries use a custom entry point. Without `termination` the `Termination` trait doesn't exist (it is defined in `std` after all).

Also it would be best if using a custom test runner didn't require an allocator as that requires quite a bit of boilerplate (e.g. `#[alloc_error_handler]`) and uses up precious Flash memory.

Of course I'm not going to block any PR since this is in eRFC state and the idea is to experiment; 
i just want to make sure that the embedded use case is not forgotten and that what we end up stabilizing does support it.

cc @Manishearth I thought `cargo fuzz` used `#![no_main]` and that it had similar needs as the embedded use case, but perhaps the fuzzed programs do link to `std` so they are not that similar?.

**Comment by djrenren** [comment]
@japaric : support for the embedded use case has been a goal for me from day one, though I will confess I'm not intimately familiar with it.

The `Testable` trait in the proposal is defined by the testing framework, not the language, so it can definitely be designed without the need for allocation.

As for the `#![no_main]` problem I'm not very familiar with it so I'll have to review a bit, but I see the issue... Might have to revise a bit so the test runner can be called from a generated `start` instead of `main`.

**Comment by Manishearth** [comment]
Yeah, cargo-fuzz uses no_main but I thought we had dropped the use case from the original eRFC? (or had considered it low priority).

cargo-fuzz can be _changed_ to be non-no_main with some libfuzzer edits.

**Comment by djrenren** [comment]
Hey all,

As I'm wrapping up my internship and will have less time to work on this, I figured I'd leave some notes for the community on the state of things merged so far and going forward. (I'll still be around, but a full-time job will definitely slow my contributions).

Current Status:
 - `#[test_case]` and `#![test_runner]` are merged
 - No cargo integration currently (though they work just fine with cargo test)


Work to be done:
 - Migrate `test::test_main_static` to accepting a trait instead of a struct so it can be compatible with other test data formats beyond `TestDescAndFn`
 - Work with various test crate authors to evaluate the ergonomics of the system. (`criterion` or `wasm-bindgen` is probably a good place to start).
 - Experiment with abstractions on top of these tools. It may turn out that the primitives are *too* primitive

Having presented this once at Mozilla and at a meetup, I've gotten a bit of feedback that's worth mentioning. There's concern about only having a single test runner for the crate. @SergioBenitez recommended having `#[test_case]` be parameterized with what kind of test it's annotating, and building separate binaries for different kinds of tests.

I think this adds a reasonably heavy requirement on configuration and build-tools so I'm not sure I agree but the point is well-taken that cooperation between tests and runners could be tenuous under the current circumstances. I believe that having a default `Testable` trait accessible in the compiler could act as defacto standard for compatibility.

I'm sure he could say more about it and offer a stronger argument, but either way this is something we should sort out. Luckily these things could be implemented on top of the current primitives so experimentation should be doable.

**Comment by SergioBenitez** [comment]
Happy to expand on that, @djrenren.

#### The Problem

The current implementation makes it impossible, without workarounds or test runner intervention, to write tests that are intended to be run by different runners in a single crate. For example, consider a crate that would like to use both `quickcheck` and `criterion`:

```rust
#[quickcheck]
fn identity(a: i32) -> bool { ... }

#[criterion_bench]
fn benchmark(c: &mut Criterion) { ... }
```

Under the current plan, this would expand to:

```rust
#[test_case]
fn identity(a: i32) -> bool { ... }

#[test_case]
fn benchmark(c: &mut Criterion) { ... }
```

To run these under the current proposal, a user would need to provide a _single runner_ that can execute both of these test cases. Of course, that's clearly not the intention here; one runner (`quickcheck`) is responsible for `identity`, while a different runner (`criterion`) is responsible for `benchmark`. Alas, there is no way to express this.

#### Solutions

I see two potential solutions to this problem: one puts the onus on the test runner and `cargo` while the other on the std. lib and `cargo`. The first requires test runners to emit a `cfg` when they emit a `test_case` to ensure that annotated items are only compiled when that runner is requested. With this approach, `criterion` and `quickcheck` would be modified so that the original code would instead expand to something like:

```rust
#[test_case]
#[cfg(target_runner = quickcheck::run)]
fn identity(a: i32) -> bool { ... }

#[test_case]
#[cfg(target_runner = criterion::bench)]
fn benchmark(c: &mut Criterion) { ... }
```

When a runner `foo::bar` is selected, `cargo` would set the `target_runner = foo::bar` flag automatically.

The second solution is almost exactly like the first but moves the responsibility for emitting the `cfg` attribute from the test runner crate to the standard-library. Under this approach, the `test_case` attribute takes a single parameter: the path to the runner intended to be used for the given test case. With this approach, `criterion` and `quickcheck` would expand the original code to something like:

```rust
#[test_case(quickcheck::run)]
fn identity(a: i32) -> bool { ... }

#[test_case(criterion::bench)]
fn benchmark(c: &mut Criterion) { ... }
```

Then, a `test_case(runner)` is only compiled when `runner` is the selected runner; _how_ this happens is an implementation detail, but I presume it would work similar to the `cfg` approach above.

#### Summary

I believe this second approach is the better approach because it is less error prone, more succinct, and hides implementation details. With the first approach, _all_ crates must follow the appropriate convention or risk breaking consumer crate: if even a single crate _does not_ `cfg` a `test_case`, it again becomes impossible to write test cases for multiple runners in a single crate. What's more, the first approach results in adding boilerplate to every test runner crate and requires `cargo` to recognize a new `cfg` that must be stabilized.

> I think this adds a reasonably heavy requirement on configuration and build-tools...

I disagree. If `cargo` is going to be aware of test runners anyway, having it emit a `cfg` that recognizes which runner was selected is a trivial addition.

**Comment by CAD97** [comment]
I like the idea of `#[test_case(quickcheck::run)]`. Just to strawman out an idea that would allow for running multiple test runners in one go:

```rust
fn main() { // entry generated by `cargo test`
    let libtest_tests = &[ /* collected tests */ ];
    let quickcheck_tests = &[ /* collected tests */ ];
    let criterion_tests = &[ /* collected tests */ ];

    ::test::run_tests(libtest_tests);
    ::quickcheck::run(quickcheck_tests);
    ::criterion::bench(criterion_tests);
}
```

`cargo-test` would allow running specific tests filtering as normal, but also allow filtering by runner if you, say, only wanted to run the `criterion` benchmarks, or wanted to exclude them from your normal test. (For that matter, maybe the same treatment should be given to `cargo-bench`, add a `#[bench_case]` attribute, so that benchmark crates don't have to pretend to be test runners.)

**Comment by djrenren** [comment]
@SergioBenitez - My rebuttal:

I think your view of test runners is different from what I envisioned and that's causing some confusion but firstly, I'd like to point out that the multiple test runner case works under the current implementation and looks roughly like so:

```rust
#![cfg_attr(quickcheck, test_runner(quickcheck::runner))]
#![cfg_attr(criterion, test_runner(criterion::runner))]


#[cfg(quickcheck)]
mod quickcheck_tests {
    #[quickcheck]
    fn identity(a: i32) -> bool { ... }
}

#[cfg(criterion)]
mod criterion_benches {
    #[criterion_bench]
    fn benchmark(c: &mut Criterion) { ... }
}
```

This misrepresents the relationship between a test and its runner though. In general, **tests should know how to run themselves**. In fact, this is how `#[test]` works today.

```rust
#[test]
fn foo(){}
```
expands to:
```rust
fn foo(){}

#[test_case]
const foo_gensym: TestDescAndFn = TestDescAndFn {
    desc: TestDesc {
        name: TestName::Static("foo"),
        ignore: false,
        should_panic: ShouldPanic::No
    },
    testfn: TestFn::StaticTestFn(|| {
         assert_test_result(foo())
    }),
}
```

Where `testfn` captures how to run the test. It enforces the interface that tests panic if they return a non-zero termination.

Test runners are concerned with structural information about the test. Anything needed for reporting, filtering, or dispatching. This is why quickcheck can reduce to a `#[test]` function with a simple `() -> ()` interface (and transitively into a `TestDescAndFn`. 

It's reasonable for a minimal test interface to look like:

```rust
trait Testable {
     // panic on failure
     fn run(&self);
     fn name(&self) -> String;
     fn is_bench(&self) -> bool;
}
```

Then  your example:

```rust
#[quickcheck]
fn identity(a: i32) -> bool { ... }

#[criterion_bench]
fn benchmark(c: &mut Criterion) { ... }
```

Would expand to:

```rust
fn identity(a: i32) -> bool { ... }

#[test_case]
const identity_test: SimpleTest = SimpleTest {
   test_fn: || { quickcheck_run(identity) },
   test_name: "identity",
   is_bench: false
}

fn benchmark(c: &mut Criterion) { ... }

#[test_case]
const benchmark_test: CriterionTest = CriterionTest {
   bench_fn: benchmark,
   test_name: "benchmark"
}
```

And there would be respective impls for the rest runner: 

```rust
impl Testable for SimpleTest {
   fn run(&self) {
      self.test_fn()
   }

  fn name(&self) -> String {
     self.name.to_string()
  }

  fn is_bench(&self) -> bool {
     self.is_bench
  }
}

impl Testable for CriterionTest {
  fn run(&self) {
    self.bench_fn(criterion::global_instance())
  }

  fn name(&self) -> String {
    self.name.to_string()
  }

  fn is_bench() -> bool {
    true
  }
```

Explictly: **there should not be a 1:1 relationship between test runners and test declaration styles**

This allows the test runner to manage the execution of tests that were not created with it in mind. 

Under the proposed addition of `#[test_case(some_runner)]` then the test declaration-style author (for example, quickcheck) would make the call of the test runner. 



> Why is the user restricted to selecting a single test runner? It would be nice to compile one binary that executes test cases from multiple runners.

A test runner is essentially a main function. It doesn't make sense to have two for the same reasons that it doesn't make sense to have two `main` functions. Because a test runner defines the command line interface for a test binary, having multiple would result in contention on this interfaces.

For example, @CAD97's suggestion is actually what I implemented at first. The issue here is that if `    ::test::run_tests` and `::quickcheck::run` allow for filtering by test name. What happens when I run:

```bash
$ ./test_bin mod_a::mod_b::test_name
```

You'd need a more complex mediator that did something like:

```bash
$ ./test_bin quickcheck mod_a::mod_b::test_name
```

If we allow multiple test runner then this mediation would have to be provided by the compiler, which puts us back into the realm of being opinionated about how test binaries should work. Such a setup could be implemented in user-space with the current implementation.

**Comment by SergioBenitez** [comment]
> In general, tests should know how to run themselves. In fact, this is how #[test] works today.

I'm not sure what you could mean by "tests should know how to run themselves". A `#[test]` clearly doesn't know how to run itself: it requires the testing library to parse things like `should_panic`, to execute the function, and to catch panics. A `#[quickcheck]` test also clearly doesn't know how to run itself: it requires `quickcheck` to run the function against a myriad of inputs and then prune to find a minimal failure, which again, are foreign to the test itself.

> Explictly: there should not be a 1:1 relationship between test runners and test declaration styles

Sure, but I don't see how my proposal enjoins this.

> Under the proposed addition of #[test_case(some_runner)] then the test declaration-style author (for example, quickcheck) would make the call of the test runner.

I think this is the desire in almost every case. I understand the principal behind this plug-and-play, swappable runner, trait-based testing, but I don't see it getting use in practice. To think about what such use would mean: one crate would propose a testing interface, and some other crate would say "Oh, sure, I can run that!". It won't happen serendipitously, since the interface will become part of the public API, so one crate is making a decision to allow other runners to run "its" tests. If that's the case, then that crate can simply take in a runner in its own attribute:

```rust
#[quickcheck(some::runner)]
```

> A test runner is essentially a main function. It doesn't make sense to have two for the same reasons that it doesn't make sense to have two main functions.

This is only true because you're defining it that way.

> Because a test runner defines the command line interface for a test binary, having multiple would result in contention on this interfaces.

I think this is an interesting design point: _should_ a test runner be able to define the CLI for the binary? I can see arguments both ways. If not, then this is a non-issue, but if so, I wouldn't be opposed to passing test-runner specific arguments, as you suggest, with something like:

```rust
$ ./test_bin --quickcheck quickcheck args here --criterion some other args
```

This would mean that either the binary can't accept any `--` args, or the top-level CLI parser is clever enough to group things appropriately.

Overall, I think your design optimizes for the wrong thing while making a common thing very difficult and tedious to accomplish. I don't see a compelling argument for swappable runners, and making it cumbersome to have tests for multiple runners in a single crate, a situation I postulate will be very common (something my code does now, for instance), is unfortunate.

**Comment by jonhoo** [comment]
I'd also like to push back a little bit here @djrenren: it's not clear to me that a "clean" division of "test runners are just main functions that deal with CLI + output" and "tests that know how to run themselves" is sufficient. It *could* be, but it'd be nice to see some evidence for it (similarly, @SergioBenitez, can you think of any examples where that separation is *not* sufficient?). There are definitely cases where a test runner should have more knowledge about the exact tests, such as "run at most this many tests in parallel within this test suite", or "this suite of tests share the following set-up, which should only be run once". It could be that these can all be dealt with using shared state among the generated tests + mutexes, but some discussion around it would be good :)

**Comment by shepmaster** [comment]
> tests for multiple runners in a single crate, a situation I postulate will be very common

Could you share some other places where this is common?

For example, my experience has not included any language where a project has multiple test runners in a single "place". The closest I can think of is multi-language projects. For example, the Rust Playground has Rust unit tests, JS unit tests, and Ruby integration tests, each with their own test runners. Thus I postulate it is *not* common 😇 

Even in a Rust-only project, I think I'd *prefer* to place benchmarks in a completely separate file (a.k.a. a separate crate) from unit tests. I never want to run tests and benchmarks together and the information reported for each is drastically different. 

>  I understand the principal behind this plug-and-play, swappable runner, trait-based testing, but I don't see it getting use in practice

I agree that it seems unlikely that there will be One True Trait for defining a test. I'd expect there to be multiple test runners, each with their own highly-specific trait. 

What I *would* expect is to see adapters for tools like proptest to connect to popular test runners, leveraging each runners strengths. Thus, the crate `proptest-stainless` would provide the macro support needed to define and connect a proptest-based test with the stainless test runner. Ideally, these macros would have high overlap with other adapters like `proptest-builtin` and `proptest-some-new-cool-runner`, but they would also allow tying into runner-specific functionality (e.g. the test output can show the number of runs or the failing inputs in a better way).

It's possible that over time some basic commonality can be discovered (perhaps something along the lines of [TAP](https://testanything.org/)?), but there will always be something special a test runner will want to provide.

**Comment by SergioBenitez** [comment]
@jonhoo I tried to give a couple of example indicative of this in my previous comment, but perhaps we have different notions of what "separation" means in this case. Could you perhaps expand on the question?

@shepmaster In Rocket, I'd love to have the following in a single crate:

  * CLI Tests
  * HTTP Client/Server Integration Tests
  * Quickcheck Type Tests
  * Fuzzers, with a time limit
  * Regular `#[test]` tests

I'm totally okay with benchmarks being in another crate. The first four are currently regular `#[test]` tests, since custom test frameworks don't exist, but I can see a nice interface for all four given this new functionality. It'd be odd to split them up into four (or five, if `#[test]` uses `#[test_case]`) crates, and having to manually specify a `cfg` to run the particular tests means not only more code, but also that with high probability, users won't run them before they submit a PR.

**Comment by mathstuf** [comment]
Apologies if this is the wrong place for this, but it seems to be the furthest along the road I've traveled in search of skipping tests by detecting support for what is being tested at runtime. Basically something like `#[ignore(lacks_support)]` where `fn lacks_support() -> bool` (or maybe `-> Result<(), String>` to give a reason support is missing) is a function within the relevant scope. My use case is that I'm testing features tightly tied to the kernel that is being run (this is for the [keyutils](https://github.com/mathstuf/rust-keyutils) crate) and I'd like to have the tests indicate that they are not actually giving useful information rather than just bailing out and saying "yep, totally works".

This was mentioned in <https://internals.rust-lang.org/t/pre-rfc-make-test-and-bench-more-flexible/1328>, but that discussion has died down.

So, is this something that is basically under this RFC impl (as something like an additional method on the trait for whether the test should be ignored or not) or is it something that should go into a new RFC? Thanks.

**Comment by shepmaster** [comment]
@mathstuf I think [Is there a simple way to conditionally enable or ignore entire test suites in Rust?](https://stackoverflow.com/q/50565893/155423) and [Is it possible to conditionally enable an attribute like `derive`?](https://stackoverflow.com/q/42551113/155423) cover your needs if it's statically determinable, like a feature flag.

**Comment by mathstuf** [comment]
Using a `build.rs` to determine it is probably sufficient for now (though the Rust bindings for performing these checks is what is being built…so that seems awkward…unless tests can have their own `build.rs`?), but I think other things also fit under this umbrella. This also has precedent in other test frameworks (e.g., pytest has [support for skipping](https://docs.pytest.org/en/latest/skipping.html) based on runtime checks).

**Comment by mathstuf** [comment]
Actually, `build.rs` would completely break for cross-compiling because what matters is the target machine's feature set, not the host machine's, so runtime-determined skipping still seems very relevant to me in the wider ecosystem.

**Comment by e-oz** [comment]
Thank you very much, now I can't compile my project using Rust 1.38.0-beta.1:

"use of unstable library feature 'test': `bench` is a part of custom test frameworks which are unstable"

**Comment by idubrov** [comment]
>In general, tests should know how to run themselves. In fact, this is how #[test] works today.

I'll give a counter-example. In https://crates.io/crates/datatest, each "test" is a template for tests which is "rendered" by the test runner, either based on scanning the filesystem for files matching the pattern or by scanning given YAML file for test cases.

So, there are no "tests" unless test runner renders them based on external data.

I could instead run all the test cases in a single "#[test]", but large part of `datatest` UX is that each individual test case is treated as a separate test (you can run it individually, for example, the same way you can run individual test with `cargo test -- <test>`).

It works well with nightly features `#[test_case]` and `#[test_runner(datatest::runner)]` (though, I have to explicitly support "regular" `#[test]`-tests). Stable is a different story... 😅

**Comment by killercup** [comment]
@idubrov, just discussed this in the dev tools meeting. As far as we could tell there was not much progress on this over the last months. We do want this to pick up speed again, though! If you'd like to help out from the rustc side too, let us know! It's always good to have actual use cases to work towards :)

@gnzlbg, it seems you worked on libtest/[extracting](https://github.com/rust-lang/libtest) libtest? Can you maybe give a quick summary of the state of this?

**Comment by idubrov** [comment]
I'm happy to help, but I'm a bit confused about the scope of the problem. I've read RFC, I have some basic understanding of other pieces, however, the big picture is something I'm still not clear about.


To me, the whole testing framework seems like a combination of the following aspects:

1. Test runners. Things like scheduling threads, capturing input and output, collecting statistics (counts of failed tests, and so on). Seems like requirements could very vary across different test runners:

a. Benchmark-like tests (criterion) would perhaps want to do all the scheduling themselves. I'm not very familiar with benchmarking, but I would expect they don't want multiple random threads unnecessarily as that might affect results?
b. Fuzzing tests could run in parallel, but they are essentially infinitely long, so they, perhaps, need their own scheduling, too?
c. "Datatest" tests want to 100% piggy-back on the standand test runner, though, as they are like regular tests, but "rendered" at runtime from test declaration pattern.
d. Not sure about "quickcheck", but maybe it falls in the same category as "datatest"?

2. API to tests execution. This would be `cargo`-based command-line API as well as IDE discovering tests (which, perhaps, should work via `cargo` anyways). Given the previous item, it seems like different tests should have different entry points. However, at the same time, if custom test kinds are allowed and, essentially, unlimited, there still should be a single entry point to the whole bag of test runners somehow? Otherwise, how would you run them? Create `cargo-xyz-test` for each test framework?

Another reason to allow everything in one crate would be so you don't have to compile same binary for each test runner. On the other hand, might be necessary for some types of tests (like, benchmarking tests requiring "release" build)?

3. Declaring tests. In my opinion, ideal scenario would be if you can mix and match tests of all various kinds in the same crate (in other words, you should not have to select only one single test runner for your crate).

My personal preference would be something like https://docs.rs/linkme/0.1.3/linkme/ or https://internals.rust-lang.org/t/idea-global-static-variables-extendable-at-compile-time/9879. I think, it would be a very nice generalization of what compiler currently achieves with `#[test_case]`.

But then again, I don't know if that would work, for example, for fuzzing tests? They might want to do compilation their own way?

4. Rendering test "instances" based on their static declaration in code. This is to take test declarations from 3 and generate input for test runners in 1.

I make this very artificial separation between test "declaration" (what you see in code) and actual test instance (what you see in the output of `cargo test -- --list`). However, this separation would server "datatest" the best. Currently I override standard test runner and "render" tests before delegating to it, but in fact, I don't really want to intercept test runner. Even more, I would love to be able to support other test runners (say, criterion) by "expanding" to their test declarations from my test "patterns".

I guess, for "datatest" in particular, this might be generalized as simple as saying that these test declarations should be lists of functions that return test declarations rather than list of declarations themselves (and then the format of these declarations, like https://doc.rust-lang.org/test/struct.TestDescAndFn.html, should be stable public API).

**Comment by djrenren** [comment]
Hey @idubrov, I can answer a lot of these questions.

### 1. Test Runners
Every executable needs a single entrypoint, so when we build a test executable there needs to be some `main` and that `main` needs access to all the tests it's going to run. This inherently couples three problems: What is a test? How do I run it? How do I report it?

The test runner exists to answer both questions. For `libtest` a test is a `TestDescAndFn`, it runs them by invoking the wrapped function, and it reports it using its built-in reporting infrastructure.

This is okay but as you've noted it leaves out a lot of possibilities, for example, this is a very flat structure, if you want test suites with setup and teardown doing it in this model is awkward. Or if you want to test without enable panic unwinding you're out of luck (this is actually still a problem with the current `test_runner` setup but it should be fixable).

So the test runner has to answer these questions, but it can delegate them to the test through the use of some trait. Imagine it had trait like `MyTestTrait` with the method `report()`. Then this test runner becomes flexible.

### 2. Command line API
In order to allow CTF's to work in all environments some of the logic should be in `rustc`. You raise a valid point that building for `test` shouldn't imply `dev`. 

Right now, `rustc` can accept a `test_runner` parameter the same way it receives any other crate attribute. By making `rustc` flexible enough, we can allow Cargo to orchestrate these options and shape convention. For this reason we left the `cargo` command-line interface mostly out of scope. But note, that `test_runner` can be used even in the Cargo integration test files.

### 3. Declaring tests
As stated above, there needs to a single entrypoint to the executable so "multiple test runners" essentially means "one super generic test runner" which you can build today. If there ends up being a right™ way for such a thing to work, we could consider stabilizing it. For now, the tools to build it exist.

The other option is compiling different executables for each test runner. This is also doable today using `#[cfg(..)]` so if you put `#[cfg(quickcheck)]` over each test that's meant to run with quickcheck and you put `#![cfg_attr(quickcheck, test_runner(quickcheck::runner)]` in your `lib.rs` then you're off to the races. The former can be hidden macros and the latter could eventually be orchestrated by Cargo profiles.

Solutions like `linkme` struggle from not being particularly cross-platform, whereas `test_case` works by manipulating the rust AST ensuring that its aggregation works everywhere `rustc` works.
When I was designing the CTF approach, people were rightfully concerned about creating a truly generic aggregation technique and opening pandora's box to all sorts of weird shenanigans in the ecosystem. Also the affects of aggregation on incremental compilation were (and probably still are?) not well understood. `#[test_case]` works well because it could eventually be reimplemented in terms of such a primitive (should it ever come to exist) in much the same way `#[test]` is now built with `#[test_case]`

### 4. Test Manifests
All that `cargo test -- --list` does is build the test executables and pass in `--test` to them. This means you could build any similar behavior into a test runner today. Case and point: libtest has it. It could be implemented for a more generic interface than `TestDescAndFn` but ultimately that's all there is to it.


P.S. While I've covered a lot of it here, my user-centric writeup of the current implementation is here: https://blog.jrenner.net/rust/testing/2018/08/06/custom-test-framework-prop.html

**Comment by idubrov** [comment]
@djrenren how this proposal relates to the whole RFC? Is it one possible direction? 

I still can't make proper connections in my head; your proposal seems to be at least partially implemented in nightly Rust (and custom test runner is what I use in my "datatest" on nightly).

However, RFC itself is very confusing.

**Comment by djrenren** [comment]
Ah yeah, the RFC is very wishy-washy (by nature of being an eRFC). It basically says something along the lines of "find use cases, make them work".

The `test_runner` approach is fully implemented in nightly rustc (though, there's some weirdness around `no_std` to work out). The current approach in nightly is one possible direction, but developed with a good deal of input from the tools team. From a design approach it represents a "maximally generic, minimal surface area" approach. In my opinion, it represents the best path forward, but at the very least, it provides a really powerful base on which to experiment with different APIs if you have different ideas.

The RFC really struggles from a lack scoping. For example, there will always be some kinds of testing we don't support, but the RFC doesn't draw this line. For my approach, I ended up ruling out doctests and compiletests, because these rely on concepts above the language the level. (Though a a variation of each could be expressed)

The purpose of the current implementation was to provide a grounding for a specific RFC (which I arguably should have written, but then grad school caught up with me). Through implementation we were able to learn a lot about what's possible, what's advisable, etc. that wasn't known when the eRFC was written. So that's the state of things.

Does that make sense?

**Comment by mathstuf** [comment]
> Ah yeah, the RFC is very wishy-washy (by nature of being an eRFC). It basically says something along the lines of "find use cases, make them work".

Does my [above comment](https://github.com/rust-lang/rust/issues/50297#issuecomment-510085386) count as something which counts as a "found use case" that can be made to work in an official RFC? Would it help if support for such a thing were added to your experimental branch?

**Comment by e-oz** [comment]
So you just don't care that you brought breaking change into language? Don't break your own promise, call it Rust 2.0 then.

**Comment by djrenren** [comment]
@mathstuf, for sure! Here's a sketch of how you'd solve your problem using the currently tools available in nightly:

1. Create a type to describe tests that allows them to report a reason for skipping.
2. Create a test macro that produces a `#[test_case] const ....` value that generates values of this type.
3. Create a test runner that accepts these and knows how to report reason for skipping.

Right now this is a little painful if you're migrating from existing libtest stuff. Ideally you'd just declare a new trait to describe your `SkippableTest` type and implement it for `TestDescAndFn` and then you'd have backwards compatibility with regular `#[test]` functions. Unfortunately, we can't do this because `TestDescAndFn` isn't stable. This is one of the reasons @gnzlbg's work on extracting libtest is so important. If `libtest` is just a regular crate you can add to your dependencies then you can rely `TestDescAndFn` from that crate which has far-weaker stability requirements than anything that ships with rustc.

**Comment by shepmaster** [comment]
@e-oz please keep discussion on-topic and civil. Nothing being discussed here involves breaking changes as far as I can see. If there *are* breaking changes, we appreciate them being pointed out as we do not wish to break backwards compatibility.

**Comment by djrenren** [comment]
@e-oz it sounds like your issue is the result of a library you consume taking advantage of unstable rust features, which is not allowed outside of nightly. This is a long-standing policy of the compiler to have nightly-only features to allow for experimentation with APIs before they are stabilized.

If `#[bench]` was working on stable or beta rust then it was a bug, and it got fixed. If this happened because you moved from nightly to beta then it is expected behavior. You can solve your problem by going back to nightly, though in general I don't recommend depending on unstable nightly features as they are subject sudden change.

**Comment by idubrov** [comment]
>Does my above comment count as something which counts as a "found use case" that can be made to work in an official RFC? Would it help if support for such a thing were added to your experimental branch?

Yes, this is something that is already supported with nightly features  / @djrenren proposal. The whole "datatest" thing I'm doing is basically generating test instances at runtime. Ignoring at runtime works, too (here I ignore by path, but as well I can ignore by anything else): https://github.com/commure/datatest/blob/114ea40c983ae108f6d7b70e6bb6fc5f3de56e0f/tests/tests/mod.rs#L74-L86

One thing, though, is that this requires overriding test runner. I wonder if lot of use-cases could be covered by allowing list of tests to be `fn() -> &[TestDescriptor]` instead of `TestDescriptor`. Or any other way (compatible with `no_std`) of allowing to generate test descriptions at runtime rather than provide them directly?

**Comment by mathstuf** [comment]
OK. If we have docs for doing that, that sounds fine to me. I really just want the test report to explicitly say "skipped because $reasons" rather than:

  - not showing up at all (compile-time decisions)
  - spurious passing (because nothing was actually tested)
  - spurious failure (really annoying)

Right now, my solution is to have a "sentinel" test whose output is basically "this module's tests are going to fail because the features aren't supported".

Thanks; I'll take a look at nightly when I get a chance to see what I can do there.

**Comment by djrenren** [comment]
Oh and just in case you aren't aware, if you can determine that you want to ignore the test at compile time, you can use the `ignore` attribute either with attr_if or your own macro to conditionally insert it.

**Comment by mathstuf** [comment]
I'm aware; it's just that for cross-compiling, compile-time detection is kind of useless for detecting kernel features available :) .

**Comment by e-oz** [comment]
> @e-oz it sounds like your issue is the result of a library you consume taking advantage of unstable rust features, which is not allowed outside of nightly. This is a long-standing policy of the compiler to have nightly-only features to allow for experimentation with APIs before they are stabilized.
> 
> If `#[bench]` was working on stable or beta rust then it was a bug, and it got fixed. If this happened because you moved from nightly to beta then it is expected behavior. You can solve your problem by going back to nightly, though in general I don't recommend depending on unstable nightly features as they are subject sudden change.

No, I'm not retarded and I understand difference between nightly and beta. I never was using nightly.

**Comment by e-oz** [comment]
> @e-oz please keep discussion on-topic and civil. Nothing being discussed here involves breaking changes as far as I can see. If there _are_ breaking changes, we appreciate them being pointed out as we do not wish to break backwards compatibility.

I quoted message from compiler.

**Comment by killercup** [comment]
> "use of unstable library feature 'test': bench is a part of custom test frameworks which are unstable"

> And it's not even inside my code (library I use have this error), so I can't change it.

> I quoted message I got from compiler.

As far as I can tell, you've not given enough details, not even the name of the library.

@e-oz, please try to report issues in a civil way. We are not here to guess what errors you have nor are we likely to spend our time on any of it when you comment in that kind of tone.

**Comment by e-oz** [comment]
@killercup you don't need to guess anything, I quoted text of the error for you.
One more time:

> use of unstable library feature 'test': bench is a part of custom test frameworks which are unstable

**Comment by e-oz** [comment]
Also, everything I wrote is pretty "civil". 0 insults so far.

If you expect me to be "happy" about not being able to compile project without rolling back to 1.37 - do not expect it. It's not just "hobby" project, it's my job, so yes I'm very worried that some day I will not be able to update Rust version just to be able to compile this binary. And it's too huge to rewrite, so no, I can't stop using that library or replace it, so name of the library doesn't matter.

**Comment by Manishearth** [comment]
@e-oz there is not nearly enough information for us to figure out what the regression is. As far as we can tell, this should not happen on code that used to compile. It's totally possible that there's a bug causing a regression, but without further details we cannot help. You have been asked for details once already and you had an unhelpful response.

Your tone is very hostile, and you are being curt and unhelpful. This will lead to moderation action if you continue down this tack.

**Comment by Manishearth** [comment]
Also, this is a tracking issue, please file a separate issue for this and link it there. Please provide instructions for reproducing the regression.

**Comment by e-oz** [comment]
@Manishearth 
> As far as we can tell, this should not happen on code that used to compile
But it does.
> but without further details we cannot help
And I should guess what details you need?
> and you had an unhelpful response
If compiler's message is unhelpful - it's not my fault.

https://play.rust-lang.org/?version=beta&mode=debug&edition=2018

>Your tone is very hostile
This bug created real issue for me and you just prefer to ignore it with response "it should work", so what tone you expect? I should be happy about this response? You've marked all my messages as offtopic, just to show how much it "matters" for you, so what you expect from me? Begging you? I don't care about your "moderation action", I clearly see you don't want to fix breaking change you brought so I have no reason to talk with you.

What I learned is I can't trust your promises and code can just stop compiling one day.

> Please provide instructions for reproducing the regression.
It took few seconds to check what is written in compiler's message:
https://play.rust-lang.org/?version=beta&mode=debug&edition=2018&gist=b006497caba8892a0350a517d49c7a2d

It compiles in stable and doesn't compile in beta mode.

**Comment by Manishearth** [comment]
> But it does.

Like I said, it's totally possible that there's a bug, but we can't figure that out without further information.

> You've marked all my messages as offtopic

Yes, I told you to file a separate issue. Further comments here will be deleted.

You have provided enough information, but please do it in an issue form.

(I'd file it myself but I'm on a train right now)

**Comment by CAD97** [comment]
I created an issue for the `#[bench]` regression at #63798. Please direct any further discussion there.

In the future, if code stops compiling between stable and beta, create a minimized example and open an issue, rather than seeking out the linked tracking issue and just posting the error without context. It's _almost always_ a bug if something stops compiling between stable and beta. Nobody will complain about an issue making the fact known. But we can't diagnose what broke without a reproduction.

**Comment by e-oz** [comment]
@CAD97 
> rather than seeking out the linked tracking issue

Look at the compiles message please, notice the URL:

```
Compiling playground v0.0.1 (/playground)
error[E0658]: use of unstable library feature 'test': `bench` is a part of custom test frameworks which are unstable
 --> src/main.rs:5:3
  |
5 | #[bench]
  |   ^^^^^
  |
  = note: for more information, see https://github.com/rust-lang/rust/issues/50297
```
> and just posting the error without context

I was able to reproduce an error just using that message. It took few seconds literally. Yes, I had no context also, it's not my code.

> But we can't diagnose what broke without a reproduction.

Come on, nobody tried, response was "it should work, get lost".

**Comment by Manishearth** [comment]
The response was "we need more information".

If you keep commenting on this issue on this topic this will lead to a temporary ban from GitHub. Do with this information what you wish.

**Comment by e-oz** [comment]
I don't give a fuck about your threats :)

**Comment by gz** [comment]
Is there a plan or way to integrate the #[should_panic] attribute with custom test runners and #[test_case] or is this out of scope for this eRFC?

**Comment by djrenren** [comment]
@gz for sure! In fact, `#[should_panic]` is already, currently implemented using custom test runners and `#[test_case]`!

Here's how it works:

First off, because we want to react to a panic, we need to compile with panics as unwinds (as opposed to aborts). This is the default behavior of the compiler, but you can specify otherwise in `Cargo.toml`. I believe the compiler forces this mode for test compilations (which it shouldn't and we should fix that).

Now, we need a test runner that knows how to react to panics. Enter [`libtest::test_main_static`](https://docs.rs/libtest/0.0.1/libtest/fn.test_main_static.html). This is the test runner that rustc uses by default. It receives each test case as a [`TestDescAndFn`](https://docs.rs/libtest/0.0.1/libtest/struct.TestDescAndFn.html). After a bit of setup, it invokes [`TestDescAndFn::testfn`](https://docs.rs/libtest/0.0.1/libtest/struct.TestDescAndFn.html?search=#structfield.testfn) using [`std::panic::catch_unwind`](https://doc.rust-lang.org/beta/std/panic/fn.catch_unwind.html) [[src](https://github.com/rust-lang/libtest/blob/dbf328db62eacebe1aa719fa806c729bafa319b3/libtest/lib.rs#L1492)], which will allow the test runner to respond to the fact that a test failed.

All that's left is informing that this test was in fact meant to panic. This is encoded within [`TestDesc`](https://docs.rs/libtest/0.0.1/libtest/struct.TestDesc.html) which is available at  `TestDescAndFn::desc`. So all we need to do is set `desc.should_fail = true` on tests that should fail.

Enter the `#[test]` macro which takes as input a simple a test function like so:
```
#[test]
fn my_test() -> {}
```

and transforms it into a `TestDescAndFn` that's annotated with `#[test_case]`.
If `#[test]` sees that that the underlying function is annotated with `#[should_panic]` then it will output a `TestDescAndFn` where `should_fail` is `true`.

**Comment by gnzlbg** [comment]
@alexcrichton simplified the build process of rust significantly recently, so retrying the split might now work - not sure. Moving compiletest out of tree is worth doing anyways, and would help. Having to deal with two slightly different compiletests is a bit of a pain.

**Comment by gnzlbg** [comment]
Thanks, part of the extraction is merging that somehow with compiletest-rs in crates.io. I'll comment on the PR.

**Comment by ligurio** [comment]
> [**Standardizing the output**](https://github.com/rust-lang/rfcs/blob/master/text/2318-custom-test-frameworks.md#standardizing-the-output-1)

> We should probably provide a crate with useful output formatters and stuff so that if test harnesses  desire, they can use the same output formatting as a regular test. This also provides a centralized  location to standardize things like json output and whatnot.

Please consider standard test output formats like Test Anything Protocol ([spec](https://testanything.org/tap-version-13-specification.html)), SubUnit ([spec](https://github.com/testing-cabal/subunit/blob/master/README.rst)) and xUnit ([spec](https://llg.cubic.org/docs/junit/)). These formats have open specifications, widely used (especially xUnit and TAP) and already supported in many test report systems and continuous integration systems.

**Comment by gilescope** [comment]
Hi, in 2017 I asked if we could have team city service messages output for running cargo tests. I just thought I'd check back and see if this is now possible? Currently there's some nasty regex in teamcity's rust plugin - I think and that could get rewritten to use the json output instead, but I wondered if it was now possible to plug in custom test reporters as well as custom test runners?

**Comment by dvtkrlbs** [comment]
Are there any progress on this ?

**Comment by danakj** [comment]
Hello, I just wanted to give the feedback (on request of @Manishearth) that I was hoping to use the custom test frameworks for hooking Rust unit tests into the [C++ Gtest framework](https://github.com/google/googletest), however I am not able to. The problem is simply that:

1. `rustc --test` still has the compiler produce its own main() function.
2. The set of #[test] functions is only known to the compiler, and it inserts them into the generated main() function.

To use the custom test framework in our project, we would need to use our own existing C++ main() functions. What we need from the compiler is the list of test functions, and their attributes, so that we could then register them all with Gtest before invoking the test framework. Likely this would have to look like writing a macro where the inputs are the functions? And the default implementation of that macro writes a main() function that invokes them all in the rusty way. For us, it would generate some "register_everything()" at the root of the crate, which we could call from main().

Instead, I am falling back on introducing static initializers which do that registration, which is unfortunate. If you're interested I am doing this work in https://bugs.chromium.org/p/chromium/issues/detail?id=1293979.

**Comment by djrenren** [comment]
Ah interesting, you want to basically build tests as a library. Ultimately, the magic of #[test] in rust is, as youve realized, is breaking lexical scope and aggregating definitions from throughout the crate. At the time I built the custom test framework experimental implementation, there was a lot of resistance to supporting this primitive directly. Basically the concerns were that we shouldn’t incentivize people to break the visibility rules of the language. Furthermore, incremental compilation was new and we weren’t sure if such a primitive, if popular in libraries, would slow down builds across the ecosystem. And so we tried to build this specialized system on top of that primitive so that it was basically only useful for tests. Several years on now, I think it’s worth re-assessing this stance. In my opinion, ctor is spookier to me than a #[aggregate] built-in. Idk who to ask about that these days.

**Comment by Manishearth** [comment]
I'll note that the original version of the test frameworks RFC was designed to be like a whole-crate proc macro, with the main RFC work being that Cargo integration would be provided but the rest would be cobbled together from other crates (likely with a published crate implementing the basic Rust testrunner with hooks).

Perhaps it's worth trying that instead? Overall I feel like a lot of the use cases for this feature will need this level of flexibility.

https://github.com/Manishearth/rfcs/blob/53ba19d267244275c98a871ce894411db530bca7/text/0000-erfc-post-build-contexts.md#procedural-macro-for-a-new-post-build-context

**Comment by djrenren** [comment]
Yeah the whole crate macro is definitely a viable option, but I’m not sure they enable much that couldn’t be accomplished by combination of aggregation and proc macros. But both seem reasonable to me.

**Comment by Manishearth** [comment]
@danakj I guess y'all could require having a `gtest_init!()` macro at the top that adds a `#![no_main]` and some initializer code? Unsure if that would work.

**Comment by danakj** [comment]
Perhaps, but it would need to be at the crate root, and need as input a list of all the test functions, file!(), line!(), and their attributes (in order to read the test suite+name). Or some custom structure derived from an attribute macro on each test function.

Just skipping main isn't enough, and maybe I am not being imaginative enough at what the gtest_init!() macro is able to do.

**Comment by djrenren** [comment]
Nah I think your intuition is right. Unless you could somehow invoke the generated main

**Comment by Manishearth** [comment]
oh, good point, yeah

**Comment by phil-opp** [comment]
Maybe something like this could work:

```rust
#![feature(custom_test_frameworks)]
#![feature(test)]

#![test_runner(crate::test_runner)]
#![reexport_test_harness_main = "test_main"]
#![cfg_attr(test, no_main)]

extern crate test;

// Normally defined in your C++ code
#[cfg(test)]
#[no_mangle]
fn main() {
    run_rust_tests();
}

// The C++ code will call this function
#[cfg(test)]
#[no_mangle]
pub extern "C" fn run_rust_tests() {
    test_main();
}

// Invoked by the generated `test_main` function 
#[cfg(test)]
fn test_runner(tests: &[&test::TestDescAndFn]) {
    println!("Running {} tests", tests.len());
    for test in tests {
        print!("{}...", test.desc.name);
        match test.testfn {
            test::TestFn::StaticTestFn(f) => f(),
            test::TestFn::StaticBenchFn(_) => todo!(),
            test::TestFn::DynTestFn(_) => todo!(),
            test::TestFn::DynBenchFn(_) => todo!(),
        }
        println!("[ok]");
    }
}

#[cfg(test)]
mod tests {
    #[test]
    fn it_works() {
        let result = 2 + 2;
        assert_eq!(result, 4);
    }
}
```

Instead of running the tests, you could also only collect the test metadata that you need in a static and report it back to your C++ code. For example, to report the number of available tests:

```rust 
static mut TEST_NAMES: Vec<String> = Vec::new();

#[no_mangle]
pub extern "C" fn run_rust_tests() -> usize {
    test_main();
    unsafe { TEST_NAMES.len() }
}

fn test_runner(tests: &[&test::TestDescAndFn]) {
    unsafe { TEST_NAMES = tests.into_iter().map(|t| t.desc.name.to_string()).collect() };
}
```

**Comment by melynx** [comment]
> Ah interesting, you want to basically build tests as a library. Ultimately, the magic of #[test] in rust is, as youve realized, is breaking lexical scope and aggregating definitions from throughout the crate. At the time I built the custom test framework experimental implementation, there was a lot of resistance to supporting this primitive directly. Basically the concerns were that we shouldn’t incentivize people to break the visibility rules of the language. Furthermore, incremental compilation was new and we weren’t sure if such a primitive, if popular in libraries, would slow down builds across the ecosystem. And so we tried to build this specialized system on top of that primitive so that it was basically only useful for tests. Several years on now, I think it’s worth re-assessing this stance. In my opinion, ctor is spookier to me than a #[aggregate] built-in. Idk who to ask about that these days.

I second this. Allowing functions to be aggregated provides for maximum flexibility.
I'm working on some no_std shared objects which have a different architecture from the runner.
I wanted to just have custom_test_frameworks expose the runner test harness "main" as an exported symbol but it seems that building test assumes that it has to be compiled as an executable.

**Comment by joshtriplett** [comment]
This experiment has been open since 2018. It doesn't look like there's been recent development activity on it.

I'm going to close this for now. If someone wants to restart this effort, one start would be a lang MCP for a general mechanism for collecting an array of items of the same type from across the program, based on attributes. That mechanism would be generally useful for various things, including tests, tracing, and other annotations.

@davidbarsky expressed some interest in working on that MCP.

**Comment by shepmaster** [comment]
>  MCP for a general mechanism for

What about replacing stdout / stderr like the built in test framework? Using something like [inventory](https://docs.rs/inventory/latest/inventory/), I've gotten good enough support for collecting an array of items for my own nascent test framework. What I don't have is a good way to prevent parallel tests from combining output.

**Comment by Manishearth** [comment]
Honestly at this point I would strongly argue for my original proposal, which was a more proc-macro-esque API that would require minimal changes to the compiler and farm out all the hard stuff to libraries. The main thing that `cargo test` gets that you can't on your own is not endpoint-collection (you can do that with a macro), it's cargo being able to figure out the correct rustc invocation.

I feel like the lack of interest in dealing with the large amount of design work to get the eRFC as currently written over the hump is a pretty major signal. Plus we have documented use cases in this thread where it's insufficient.

I don't have time to go through this again, but if someone is interested I'm happy to help out. I imagine we would still need an MCP but the actual changes would be largely on the cargo side.

**Comment by Ezrashaw** [comment]
@joshtriplett I'm interested in restarting this proposal, do you have any advice?

**Comment by Manishearth** [comment]
Not much more than go through the RFC and the current code (and any PRs linked to this issue) and see what's missing. I don't remember the status quo unfortunately.

**Comment by gilescope** [comment]
I think the biggest change in rust testing has been the arrival of nextest
on the scene.

**Comment by asomers** [comment]
> I think the biggest change in rust testing has been the arrival of nextest on the scene.

cargo-nextest really doesn't address this issue.  It's just a different runner that can execute the same test.  But it doesn't allow you to alter the design of the tests themselves, for example by creating test functions at runtime, or deciding at runtime that a particular test function should be skipped.

**Comment by Ezrashaw** [comment]
Yes, although I think @gilescope is talking about changes to the Rust testing ecosystem generally.

**Comment by amab8901** [comment]
if this issue is closed, then maybe this (https://doc.rust-lang.org/core/prelude/v1/macro.test_case.html) should no longer be labeled as "nightly" and "experimental"?

**Comment by mzabaluev** [comment]
> if this issue is closed, then maybe this (https://doc.rust-lang.org/core/prelude/v1/macro.test_case.html) should no longer be labeled as "nightly" and "experimental"?

[Some crates](https://crates.io/crates/test-case) already exploit the `test_case` attribute in stable Rust, so the effects need to be documented at least.

**Comment by ctz** [comment]
I was looking at using [datatest](https://crates.io/crates/datatest) -- it seems to address my needs to produce test cases from json. However, it requires `custom_test_frameworks` which is nightly-only.

And this being the tracking RFC for that feature, what does it mean for it to be closed? Is this effectively nightly-only, forever? Will it be removed? Is it destined for the same purgatory as `#[bench]`?

If this is permanently abandoned, it would be good if this issue, and the unstable book recorded that in a clear way.

**Comment by jdonszelmann** [comment]
I just wanted to link this tracking issue here, as it's somewhat related:
https://github.com/rust-lang/rust/issues/125119

**Comment by HernandoR** [comment]
I followed unstable book of `test::Bencher` to here, after browsing dozens of issues, may i conclude that `crate test` is unstable, hence we should open to discussion here? also, what is unstable and needs to be discussed and decided to stabilize test lib?

**Comment by weihanglo** [comment]
See https://github.com/rust-lang/testing-devex-team/issues/2

**Comment by RalfJung** [comment]
Reopening, since this issue is still referenced as tracking issue by compiler messages (e.g. [here](https://github.com/rust-lang/rust/blob/22998f078588cf479530ff93e4a66ab4bb666398/library/core/src/macros/mod.rs#L1682-L1685)).

**Comment by mrchantey** [comment]
> There is currently no plan/RFC for stabilizing custom test frameworks.

What exactly does this mean for custom framework authors?

Can we rely on this feature's continued support but only on nightly behind a flag, or is there a chance it will be removed in the future?

**Comment by Manishearth** [comment]
There are no initiatives to remove it either. You cannot _rely_ on it continuing to exist but it probably will.

People using this feature should ideally work together on stabilizing it.

**Comment by mrchantey** [comment]
> People using this feature should ideally work together on stabilizing it.

Sounds good, im making great progress rewriting the [sweet crate](https://crates.io/crates/sweet) on top of `custom_test_framework`, ill look into the stabilization process if the approach gains momentum.

**Comment by mathstuf** [comment]
Is there a place to congregate for those interested in stabilizing it (my interest is in supporting [skippable tests](https://internals.rust-lang.org/t/pre-rfc-skippable-tests/14611))? `wg-testing` on Discord perhaps? Or somewhere on the community Discord? Zulip?

**Comment by Manishearth** [comment]
Yes, wg-testing.

**Comment by weihanglo** [comment]
https://rust-lang.zulipchat.com/#narrow/stream/404371-t-testing-devex

t-testing-devex as well

**Comment by mrchantey** [comment]
## Custom Doctests

A few above comments for context

Original RFC
> Another argument for punting on doctests is that they are intended to demonstrate code that the user of a library would write. They're there to document how something should be used, and it then makes somewhat less sense to have different "ways" of running them.

From @djrenren
> For my approach, I ended up ruling out doctests and compiletests, because these rely on concepts above the language the level. (Though a a variation of each could be expressed)


I really hope doctests will also make the cut, this RFC has been an absolute gamechanger for my implementation of the `sweet` test framework, allowing customization of both the runner (ie running in wasm with deno) and the output (my personal preference is the jest flavor) while still using the standard `#[test]` attribute.

![Image](https://github.com/user-attachments/assets/7e2500c1-5311-4ed1-9ae9-d0a46523f0e9)

At this stage vanilla doctests are the biggest painpoint of `sweet`. I understand its far more complex than customizing `#[test]` but if its possible I think its worth it. In more adventurous moods I've considered extracting the doctest tokens and attempting to run them via codegen but I'm sure that will present a bunch of edge cases and inconsistencies.

I've never looked into rust-lang internals but am willing to have a dig if this is actually achievable and has community support. I think the ability to run wasm doctests alone validates this feature.
